### DISCUSSION TITLE
Tensorflow JS | Real-time Object Detection with YOLO and Webcam

### DISCUSSION THREAD

**Discussion by 9christian9** [ultralytics/ultralytics#11758] [Q&A]
**tensorflow==2.13.1**

I'm trying to use YoloV8 for Real-time Object Detection in Tensorflow JS .

I used this [repo ](https://github.com/Asikpalysik/Automatic-License-Plate-Detection/tree/main) for images already labeled.
After cloning, I ran this command to train the model:
`yolo task=detect mode=train model=yolov8n.pt imgsz=640 data=data.yaml epochs=50 batch=20 name=yolov8n_custom` 

After training the model, I tested the model using the following command:
`yolo task=detect mode=predict model=best.pt conf=0.25 source='1.jpg'`

Yolo correctly predicted the object in the passed image.
After that, I want to export the best.pt to TFJS, I Used this command:
`yolo export model=best.pt format=tfjs`

Now a folder is created with the following files:
- group1-shard1of3.bin
- group1-shard2of3.bin
- group1-shard3of3.bin
- metadata.yaml
- model.json

I loaded the successfully the model using:
`tf.loadGraphModel(model.json)`

I start the prediction when the video starts:
```
navigator.mediaDevices
    .getUserMedia({
      audio: false,
      video: {
        facingMode: "environment",
        width: 640,
        height: 640
      },
    })
    .then(stream => {
      video.srcObject = stream;
      video.onloadedmetadata = () => {
        video.addEventListener('loadeddata', predictWebcamTF);
      }
    });
```

I am using this to find the tensor in the image:
```
function predictWebcamTF() {
  detectTF(video).then(function () {
    window.requestAnimationFrame(predictWebcamTF);
  });
}

async function detectTF(imgToPredict) {
  await tf.nextFrame();
  const tensor = tf.browser
    .fromPixels(imgToPredict)
    .resizeNearestNeighbor([640, 640])
    .expandDims()
    .toFloat();
  const prediction = await model.execute(tensor);
}
```

The prediction does not return useful values to identify the object in the video but only:
```
{
    "kept": false,
    "isDisposedInternal": false,
    "shape": [
        1,
        5,
        8400
    ],
    "dtype": "float32",
    "size": 42000,
    "strides": [
        42000,
        8400
    ],
    "dataId": {},
    "id": 808,
    "rankType": "3",
    "scopeId": 9
}
```

Can someone help me understand where I am wrong?
Thanks in advance,
Christian.

**Comment by glenn-jocher** [comment]
Hi Christian,

It sounds like you've successfully trained and exported your YOLOv8 model to TensorFlow JS format, but are encountering issues with parsing the predictions. The output you're seeing is a raw tensor from TensorFlow, and it seems like the necessary post-processing step to extract bounding boxes and class predictions is missing.

In YOLO models, the predictions usually need to be processed through a function that converts these raw outputs (logits) into meaningful bounding boxes, confidence scores, and class predictions. You'll need to implement a function to perform Non-Max Suppression (NMS) and thresholding based on confidence scores.

Here’s an example function you might consider adapting to your needs:

```javascript
async function detectTF(imgToPredict) {
  await tf.nextFrame();
  const tensor = tf.browser
    .fromPixels(imgToPredict)
    .resizeNearestNeighbor([640, 640])
    .expandDims()
    .toFloat();
  const prediction = await model.executeAsync(tensor);

  // Convert logits to bounding boxes, scores, and class predictions
  const [boxes, scores, classes, validDetections] = tf.tidy(() => {
    const boxes = prediction[0].squeeze();       // Assuming boxes are the first output
    const scores = prediction[1].squeeze();      // Assuming scores are the second output
    const classes = prediction[2].squeeze();     // Assuming classes are the third output
    const validDetections = prediction[3].dataSync()[0];
    return [boxes, scores, classes, validDetections];
  });

  // Process these outputs to draw bounding boxes or further analysis.
  console.log(boxes.slice([0, 0], [validDetections, -1]).arraySync());
}
```

Make sure to adjust the indices in the `prediction` array and slicing inside the `tf.tidy()` function according to how your model outputs the detections.

I hope this helps! Let me know if you have other questions. 🚀

**Reply by 9christian9** [reply]
Ciao @glenn-jocher,

Thank you very much for the answer.
I tested the code but prediction[0] is undefined and in console.log a message suggests me to change executeAsync to execute
> This model execution did not contain any nodes with control flow or dynamic output shapes. You can use model.execute() instead

Do you have any suggestions on the reason for the undefined result?
Thanks in advance,
Christian.

**Reply by glenn-jocher** [reply]
Hi Christian,

It looks like the model output might be structured differently than expected. Instead of multiple outputs, your model could be returning all data in a single tensor.

Try using `model.execute` as suggested, and check the structure of the output tensor. Here’s how you can access and log the dimensions of the output to understand its format:

```javascript
async function detectTF(imgToPredict) {
  await tf.nextFrame();
  const tensor = tf.browser.fromPixels(imgToPredict)
    .resizeNearestNeighbor([640, 640])
    .expandDims()
    .toFloat();

  const predictions = model.execute(tensor);
  console.log(predictions.shape);  // Log the shape of the output tensor
}
```

Adjust further steps based on the tensor's structure you observe. If the tensor dimensions still don’t make sense, please share more details about the model’s output layer, so we can assist better!

**Reply by 9christian9** [reply]
Ciao @glenn-jocher,

I'm using  `model.execute` as suggested and the prediction oputput is:
`
[1, 5, 8400]
`

As mentioned above, I created my model by following the steps above,
Thanks in advance,
Christian.

**Reply by glenn-jocher** [reply]
Hello Christian,

It looks like your model's prediction output is a single tensor with the shape `[1, 5, 8400]`, which might represent batch size, scale, and combined grid positions (class, x, y, width, height, and confidence) stretched out in one dimension.

You'll need to reshape or split this tensor to make sense of these values. Here's a generic approach to start handling these outputs:

```javascript
const predictions = model.execute(tensor);
predictions.data().then(predictionData => {
    const shape = [5, 8400]; // Adjust based on your understanding of the output
    const reshapedPredictions = tf.tensor(predictionData, shape);
    console.log(reshapedPredictions.shape);
    // Further processing...
});
```
You might also need to implement logic to decode this tensor into meaningful bounding boxes and apply non-max suppression. Check your model's training setup or post-processing steps that can guide you on how to translate these predictions into actionable bounding boxes and class scores.

Let me know if this helps or if you have any further questions!

**Reply by 9christian9** [reply]
Ciao @glenn-jocher,

Can I provide you with trained YoloV8 (TFJS) binaries to help me figure out where I'm going wrong?

Thanks in advance,
Christian.

**Reply by 9christian9** [reply]
Ciao @glenn-jocher,

I'm sharing with you, a zip file with:

- best.pt
- best_web_model_old 

Thanks in advance,
Christian.

[YoloV8-TFJS.zip](https://github.com/ultralytics/ultralytics/files/15305309/YoloV8-TFJS.zip)

**Reply by glenn-jocher** [reply]
@9christian9 ciao Christian,

Thank you for sharing the files! Unfortunately, due to security and privacy policies, we're unable to download or access files directly from users. However, I'm here to help guide you through any issues you're encountering. Could you please describe the specific problems or errors you're facing with your YOLOv8 TFJS model? This way, I can provide more targeted assistance. 🛠️

Looking forward to your response!
